File: agentu/app/agents.js
export function createAgentUI(d) {
  const {
    S,
    h,
    can,
    badge,
    heading,
    empty,
    actionButton: button,
    field,
    select,
    reason,
    jsonDetails,
    money,
    date,
    openForm,
    minor,
    command,
    notice,
    refresh,
  } = d;
  const providers = {
    treasury_rule: "Treasury rule",
    bedrock: "Amazon Bedrock",
    external: "External agent",
  };
  const name = (id) => S.accounts.find((a) => a.id === id)?.name || id;
  const eligible = () =>
    S.agentCatalogue.filter(
      (a) =>
        a.status === "active" &&
        a.config.provider !== "external" &&
        (a.config.provider !== "bedrock" ||
          S.capabilities?.bedrock?.configured),
    );
  const inspect = (id) => S.agentCatalogue.find((a) => a.id === id);
  const manager = () =>
    ["owner", "administrator"].includes(S.overview.membership.role);
  function credentials(a) {
    const keys = S.agentKeys.filter((k) => k.agent_id === a.id);
    return `<details class="detail"><summary>Credentials (${keys.length})</summary><p class="muted">Credentials are shown once, expire automatically and grant proposal access within this mandate. Issuing a new mandate invalidates earlier credentials.</p>${
      keys
        .map((k) => {
          const status =
            k.status === "active"
              ? k.expires_at * 1000 <= Date.now()
                ? "expired"
                : k.revision !== a.revision
                  ? "superseded"
                  : "active"
              : k.status;
          return `<div class="list-row"><div><strong>${h(k.label)}</strong><p><small>Expires ${h(date(k.expires_at))}</small></p></div><div class="actions">${badge(status)}${status === "active" && can("agent_key_revoke") ? button("agent-key-revoke", "Revoke", k.id, "danger") : ""}</div></div>`;
        })
        .join("") || '<p class="muted">No credentials issued.</p>'
    }</details>`;
  }
  function renderAgents() {
    const enabled = S.capabilities?.bedrock?.configured;
    return (
      heading(
        "Governed automation",
        "Agents & mandates",
        "Give each agent a purpose, a bounded account scope and independently approved authority.",
        can("agent_create")
          ? button("agent-create", "Create agent", "", "primary")
          : "",
      ) +
      `<div class="card capability-strip"><div><strong>Treasury rule</strong><p class="muted">Deterministic reserve rebalancing. No language model.</p></div><div><strong>Amazon Bedrock</strong><p class="muted">${enabled ? "Model configured. Inspect completed run evidence to verify execution." : "Model connection pending in the business AWS account."}</p></div><div><strong>External agents</strong><p class="muted">Scoped, expiring credentials for the proposal API.</p></div></div>` +
      (S.records.length
        ? `<div class="list">${S.records.map((a) => `<article class="card agent-card"><header><div><span class="eyebrow">${h(providers[a.config.provider])}</span><h2>${h(a.name)}</h2><p>${h(a.config.objective)}</p></div>${badge(a.status)}</header><div class="grid detail"><div><span class="eyebrow">From → To</span><p>${h(a.config.source_ids.map(name).join(", "))}<br>→ ${h(a.config.destination_ids.map(name).join(", "))}</p></div><div><span class="eyebrow">Transaction / daily limit</span><h3>${h((a.config.transaction_limit / 100).toLocaleString("en-GB"))} / ${h((a.config.daily_limit / 100).toLocaleString("en-GB"))}</h3><small>Separately in ${h(a.config.currencies.join(", "))}</small></div><div><span class="eyebrow">Review / run budget</span><h3>${a.config.require_review ? "Always review" : "Institution policy"}</h3><small>Up to ${a.config.daily_runs} runs per UTC day</small></div></div>${a.config.provider === "treasury_rule" ? `<p class="hint">Maintain a source floor of ${h(money(a.config.source_floor, S.accounts.find((x) => x.id === a.config.source_ids[0])?.currency))} and a reserve target of ${h(money(a.config.destination_target, S.accounts.find((x) => x.id === a.config.destination_ids[0])?.currency))}.</p>` : ""}<div class="actions">${a.status === "draft" && can("agent_publish") && a.created_by !== S.user.sub ? button("agent-publish", "Publish mandate", a.id, "primary") : ""}${a.status === "active" && can("agent_run") && eligible().some((x) => x.id === a.id) ? button("agent-run", "Run agent", a.id, "primary") : ""}${can("agent_revision") ? button("agent-revision", "Draft revision", a.id) : ""}${a.status !== "suspended" && can("agent_suspend") ? button("agent-suspend", "Suspend", a.id, "danger") : ""}${a.status === "active" && a.config.provider === "external" && can("agent_key_create") ? button("agent-key-create", "Issue credential", a.id) : ""}</div>${a.status === "draft" && a.created_by === S.user.sub ? '<p class="hint">A different owner or administrator must publish this mandate.</p>' : ""}${a.config.provider === "external" ? credentials(a) : ""}${jsonDetails(a, "Mandate version and publication record")}</article>`).join("")}</div>`
        : empty(
            "No agents yet",
            "Create a mandate, then have a different administrator publish it.",
          ))
    );
  }
  function renderRuns() {
    return (
      heading(
        "Execution history",
        "Agent runs",
        "Follow each request from its account snapshot through the decision and resulting operation.",
        can("agent_run") && eligible().length
          ? button("agent-run", "Start run", "", "primary")
          : "",
      ) +
      '<p class="hint">Active runs refresh automatically. Runs can propose an internal transfer or take no action. The institution policy and agent mandate govern every proposal.</p>' +
      (S.records.length
        ? `<div class="list">${[...S.records]
            .sort((a, b) => b.created_at.localeCompare(a.created_at))
            .map(
              (r) =>
                `<article class="card"><header><div><span class="eyebrow">${h(providers[r.provider])} · ${h(date(r.created_at))}</span><h2>${h(r.agent_name)}</h2><p>${h(r.instruction || r.output?.purpose || "External proposal")}</p></div>${badge(r.status)}</header>${r.output ? `<div class="detail"><strong>${r.output.action === "no_action" ? "No transfer needed" : r.output.amount ? h(money(r.output.amount, S.accounts.find((a) => a.id === r.output.source_id)?.currency)) + " → " + h(name(r.output.destination_id)) : "Decision recorded"}</strong><p>${h(r.output.reason || r.output.purpose || "")}</p>${r.decision && r.decision !== "no_action" ? `<small>Decision when proposed: </small>${badge(r.decision)}` : ""}</div>` : ""}${r.error ? `<p class="notice error">${h(r.error.message)}</p>` : ""}<div class="actions detail">${r.action_id ? button("agent-operation", "Open operations", r.action_id) : ""}${["queued", "running"].includes(r.status) && can("agent_run_cancel") && (r.requested_by === S.user.sub || manager()) ? button("agent-run-cancel", "Cancel run", r.id, "danger") : ""}</div>${jsonDetails(r, "Run evidence and identifiers")}</article>`,
            )
            .join("")}</div>`
        : empty(
            "No runs yet",
            "Publish an agent mandate to begin. External agents submit through the proposal API.",
          ))
    );
  }
  function scope(name, label, selected) {
    return `<label class="field">${h(label)}<select name="${name}" multiple size="4" required>${S.accounts
      .filter((a) => a.kind === "asset" && a.status === "active")
      .map(
        (a) =>
          `<option value="${h(a.id)}" ${selected.includes(a.id) ? "selected" : ""}>${h(a.name)} · ${h(a.currency)}</option>`,
      )
      .join(
        "",
      )}</select><small>Select up to ten. A treasury rule uses one account on each side.</small></label>`;
  }
  function edit(agent) {
    if (
      S.accounts.filter((a) => a.kind === "asset" && a.status === "active")
        .length < 2
    )
      throw new Error(
        "Create at least two active asset accounts before drafting an agent.",
      );
    const config = agent?.config || {
      provider: "treasury_rule",
      objective:
        "Maintain a treasury reserve while preserving operating liquidity.",
      transaction_limit: 7500000,
      daily_limit: 25000000,
      require_review: true,
      daily_runs: 20,
      source_ids: [],
      destination_ids: [],
      currencies: [S.overview.institution.base_currency],
      source_floor: 100000000,
      destination_target: 15000000,
    };
    const inputMoney = (key, label) =>
      field(
        key,
        label,
        "text",
        (config[key] || 0) / 100,
        "In currency units, up to two decimal places.",
        'inputmode="decimal"',
      );
    openForm(
      agent ? "Draft agent revision" : "Create agent mandate",
      agent
        ? "A new draft suspends the current mandate immediately. A different administrator must publish it; old credentials become invalid."
        : "Define explicit authority. A different owner or administrator must publish this draft.",
      (!agent
        ? field(
            "name",
            "Agent name",
            "text",
            "",
            "",
            'minlength="2" maxlength="100"',
          )
        : "") +
        select(
          "provider",
          "Decision provider",
          Object.entries(providers),
          config.provider,
        ) +
        `<label class="field">Objective<textarea name="objective" required minlength="10" maxlength="1000">${h(config.objective)}</textarea></label>` +
        `<div class="split">${scope("source_ids", "Source accounts", config.source_ids)}${scope("destination_ids", "Destination accounts", config.destination_ids)}</div>` +
        `<fieldset class="choice-group"><legend>Permitted currencies</legend>${["GBP", "EUR", "USD"].map((c) => `<label><input type="checkbox" name="currencies" value="${c}" ${config.currencies.includes(c) ? "checked" : ""}> ${c}</label>`).join("")}</fieldset>` +
        `<div class="split">${inputMoney("transaction_limit", "Transaction limit")}${inputMoney("daily_limit", "Daily limit")}</div>` +
        select(
          "require_review",
          "Independent review",
          [
            ["true", "Always require an independent reviewer"],
            ["false", "Use institution policy approval thresholds"],
          ],
          String(config.require_review),
        ) +
        field(
          "daily_runs",
          "Daily run budget",
          "number",
          config.daily_runs,
          "Counts requests, including no-action and blocked decisions, per UTC day.",
          'min="1" max="1000" step="1"',
        ) +
        `<fieldset id="rule-fields"><legend>Reserve rule</legend><div class="split">${inputMoney("source_floor", "Source balance floor")}${inputMoney("destination_target", "Destination target")}</div></fieldset>`,
      (fd) => {
        const c = {
          provider: fd.get("provider"),
          objective: fd.get("objective"),
          source_ids: fd.getAll("source_ids"),
          destination_ids: fd.getAll("destination_ids"),
          currencies: fd.getAll("currencies"),
          transaction_limit: minor(fd.get("transaction_limit")),
          daily_limit: minor(fd.get("daily_limit")),
          require_review: fd.get("require_review") === "true",
          daily_runs: Number(fd.get("daily_runs")),
        };
        if (c.provider === "treasury_rule")
          Object.assign(c, {
            source_floor: minor(fd.get("source_floor")),
            destination_target: minor(fd.get("destination_target")),
          });
        return command(agent ? "agent_revision" : "agent_create", {
          ...(agent ? { agent_id: agent.id } : { name: fd.get("name") }),
          config: c,
        });
      },
      "Save draft",
    );
    const provider = document.querySelector('[name="provider"]'),
      rule = document.querySelector("#rule-fields");
    const toggle = () => {
      rule.hidden = provider.value !== "treasury_rule";
      rule.disabled = rule.hidden;
    };
    provider.addEventListener("change", toggle);
    toggle();
  }
  function showCredential(result) {
    if (!result.agent_token)
      return notice(
        "This credential was already issued. Revoke it and issue a new one if its value was lost.",
      );
    openForm(
      "Agent credential issued",
      "Copy this credential into your agent's secret store. It is shown once and must be sent only to this Agentu environment.",
      `<label class="field">Credential<input id="agent-token" class="invite-link" type="password" autocomplete="off" readonly value="${h(result.agent_token)}"></label><button type="button" data-action="agent-copy-token">Copy credential</button><p class="hint">POST /api/agent/proposals · Authorization: Bearer &lt;credential&gt;</p>`,
      null,
    );
    document.querySelector("#dialog-submit").hidden = true;
  }
  async function act(action, id) {
    if (action === "agent-copy-token") {
      await navigator.clipboard.writeText(
        document.querySelector("#agent-token").value,
      );
      document.querySelector("#dialog-intro").textContent =
        "Credential copied. Store it securely; it is shown only once.";
      return;
    }
    if (action === "agent-create") return edit();
    if (action === "agent-revision") return edit(inspect(id));
    if (action === "agent-publish" || action === "agent-suspend") {
      const publish = action === "agent-publish";
      return openForm(
        publish ? "Publish agent mandate" : "Suspend agent",
        inspect(id).name,
        reason(publish ? "Publication reason" : "Suspension reason"),
        (fd) =>
          command(publish ? "agent_publish" : "agent_suspend", {
            agent_id: id,
            reason: fd.get("reason"),
          }),
        publish ? "Publish mandate" : "Suspend agent",
      );
    }
    if (action === "agent-key-create")
      return openForm(
        "Issue agent credential",
        inspect(id).name,
        field(
          "label",
          "Credential label",
          "text",
          "",
          "",
          'minlength="2" maxlength="100"',
        ) +
          field(
            "days",
            "Lifetime in days",
            "number",
            7,
            "Maximum 90 days.",
            'min="1" max="90" step="1"',
          ),
        (fd) =>
          command("agent_key_create", {
            agent_id: id,
            label: fd.get("label"),
            days: Number(fd.get("days")),
          }),
        "Issue credential",
        showCredential,
      );
    if (action === "agent-key-revoke") {
      const key = S.agentKeys.find((k) => k.id === id);
      return openForm(
        "Revoke credential",
        key.label,
        reason("Revocation reason"),
        (fd) =>
          command("agent_key_revoke", {
            agent_id: key.agent_id,
            key_id: id,
            reason: fd.get("reason"),
          }),
        "Revoke credential",
      );
    }
    if (action === "agent-run") {
      const choices = eligible();
      if (!choices.length)
        throw new Error(
          "Publish a configured agent mandate before requesting a run.",
        );
      const selected = inspect(id) || choices[0];
      openForm(
        "Start agent run",
        "The worker reads current scoped accounts, records its decision and applies all current controls. The person requesting this run cannot approve its resulting transfer.",
        select(
          "agent_id",
          "Agent",
          choices.map((a) => [a.id, a.name]),
          selected.id,
        ) +
          `<label class="field">Run instruction<textarea name="instruction" minlength="10" maxlength="1000" required>${h(selected.config.objective)}</textarea></label>`,
        (fd) => command("agent_run", Object.fromEntries(fd)),
        "Start run",
        async () => {
          S.page = "runs";
          try {
            await refresh();
          } catch (e) {
            notice(e.message, true);
          }
        },
      );
      document
        .querySelector('[name="agent_id"]')
        .addEventListener("change", (e) => {
          document.querySelector('[name="instruction"]').value = inspect(
            e.target.value,
          ).config.objective;
        });
      return;
    }
    if (action === "agent-run-cancel")
      return openForm(
        "Cancel agent run",
        "A late worker result will be ignored. Completed operations must be managed separately.",
        reason("Cancellation reason"),
        (fd) =>
          command("agent_run_cancel", { run_id: id, reason: fd.get("reason") }),
        "Cancel run",
      );
    if (action === "agent-operation") {
      S.page = "actions";
      await refresh();
    }
  }
  return { renderAgents, renderRuns, act };
}

File: agentu/backend/platform_core/agents.py
"""Governed machine identities and bounded agent runs.

Machine credentials never authenticate to the human administration API. They can
submit one tightly validated internal-transfer proposal, within their own mandate.
"""
import hashlib
import os
import re
import secrets
import time
from datetime import datetime, timezone
from .errors import PlatformError
from .jobs import enqueue
from .model import Actor, amount, currency, digest, identifier, new_id, now, text
from .service import PlatformService, PERMISSIONS, COLLECTIONS, MANAGERS, access, audit, required, tenant_key


def credential_key(agent_id, key_id):
    return "AGENTKEY#" + agent_id + "#" + key_id


def validate_config(tx, pk, body):
    if not isinstance(body, dict):
        raise PlatformError("Agent configuration must be an object.")
    config = {"objective": text(body.get("objective"), "Agent objective", 10, 1000),
              "transaction_limit": amount(body.get("transaction_limit"), "Agent transaction limit"),
              "daily_limit": amount(body.get("daily_limit"), "Agent daily limit"),
              "require_review": body.get("require_review", True), "daily_runs": body.get("daily_runs", 20),
              "provider": body.get("provider", "treasury_rule")}
    if config["provider"] not in ("external", "treasury_rule", "bedrock"):
        raise PlatformError("Choose external, treasury_rule or bedrock.")
    if type(config["require_review"]) is not bool or type(config["daily_runs"]) is not int or not 1 <= config["daily_runs"] <= 1000:
        raise PlatformError("Set an explicit review requirement and 1–1,000 daily runs.")
    if config["transaction_limit"] > config["daily_limit"]:
        raise PlatformError("Agent transaction limit must not exceed its daily limit.")
    for field in ("source_ids", "destination_ids"):
        values = body.get(field)
        if not isinstance(values, list) or not 1 <= len(values) <= 10:
            raise PlatformError("Select one to ten accounts for each side of the mandate.")
        config[field] = [identifier(value, "Account") for value in values]
        if len(set(values)) != len(values):
            raise PlatformError("Account scopes must not contain duplicates.")
        for value in values:
            account = required(tx, pk, "ACCOUNT#" + value)
            if account["kind"] != "asset" or account["status"] != "active":
                raise PlatformError("Agent scopes must contain active asset accounts.")
    codes = body.get("currencies")
    if not isinstance(codes, list) or not 1 <= len(codes) <= 3:
        raise PlatformError("Select one to three currencies.")
    config["currencies"] = [currency(code) for code in codes]
    if len(set(codes)) != len(codes):
        raise PlatformError("Currencies must be distinct.")
    if config["provider"] == "treasury_rule":
        if len(config["source_ids"]) != 1 or len(config["destination_ids"]) != 1 or config["source_ids"] == config["destination_ids"]:
            raise PlatformError("A treasury rule requires one source and a different destination.")
        config["source_floor"] = amount(body.get("source_floor", 0), "Source balance floor", 0)
        config["destination_target"] = amount(body.get("destination_target"), "Destination target")
        source = required(tx, pk, "ACCOUNT#" + config["source_ids"][0])
        destination = required(tx, pk, "ACCOUNT#" + config["destination_ids"][0])
        if source["currency"] != destination["currency"] or source["currency"] not in config["currencies"]:
            raise PlatformError("Treasury rule accounts must use the same permitted currency.")
    return config


class AgentService(PlatformService):
    permissions = {**PERMISSIONS, "agent_create": MANAGERS, "agent_revision": MANAGERS, "agent_publish": MANAGERS,
                   "agent_suspend": MANAGERS, "agent_key_create": {"owner"}, "agent_key_revoke": {"owner"},
                   "agent_run": {"owner", "operator"}, "agent_run_cancel": {"owner", "administrator", "operator"}}
    collections = {**COLLECTIONS, "agents": "AGENT#", "agent_keys": "AGENTKEY#", "runs": "RUN#"}

    @staticmethod
    def capabilities():
        return {"treasury_rule": {"configured": True, "uses_model": False}, "external": {"configured": True},
                "bedrock": {"configured": bool(os.getenv("BEDROCK_MODEL_ARN")), "connection_verified": False}}

    def _agent_create(self, tx, pk, tenant, actor, member, body):
        agent = {"id": new_id(), "name": text(body.get("name"), "Agent name", 2, 100), "status": "draft",
                 "revision": new_id(), "created_by": actor.sub, "created_at": now(),
                 "config": validate_config(tx, pk, body.get("config"))}
        tx.put(pk, "AGENT#" + agent["id"], agent, insert_only=True)
        audit(tx, pk, actor, "agent_drafted", agent)
        return {"agent": agent}

    def _agent_revision(self, tx, pk, tenant, actor, member, body):
        agent = required(tx, pk, "AGENT#" + identifier(body.get("agent_id")))
        previous = agent["revision"]
        agent.update(config=validate_config(tx, pk, body.get("config")), revision=new_id(), status="draft",
                     created_by=actor.sub, created_at=now())
        for field in ("published_by", "published_at"):
            agent.pop(field, None)
        tx.put(pk, "AGENT#" + agent["id"], agent)
        machine = tx.get(pk, "MEMBER#agent_" + agent["id"])
        if machine:
            tx.put(pk, "MEMBER#agent_" + agent["id"], {**machine, "status": "suspended"})
        audit(tx, pk, actor, "agent_revision_drafted", {"agent": agent, "previous_revision": previous})
        return {"agent": agent}

    def _agent_publish(self, tx, pk, tenant, actor, member, body):
        agent = required(tx, pk, "AGENT#" + identifier(body.get("agent_id")))
        if agent["status"] != "draft":
            raise PlatformError("Publish a draft mandate. To restore a suspended agent, draft and review a new version.", 409, "invalid_state")
        if actor.sub == agent["created_by"]:
            raise PlatformError("A different owner or administrator must publish this agent mandate.", 403, "independent_review_required")
        author = required(tx, pk, "MEMBER#" + agent["created_by"])
        if author["status"] != "active" or author["role"] not in MANAGERS:
            raise PlatformError("The mandate author is no longer authorised.", 409, "author_inactive")
        validate_config(tx, pk, agent["config"])
        policy = required(tx, pk, "POLICY#" + tenant["policy_id"])["config"]
        if agent["config"]["transaction_limit"] > policy["transaction_limit"] or agent["config"]["daily_limit"] > policy["daily_limit"] or not set(agent["config"]["currencies"]) <= set(policy["currencies"]):
            raise PlatformError("The agent mandate must fit within the institution's current limits and currencies.")
        agent.update(status="active", published_by=actor.sub, published_at=now(), publication_reason=text(body.get("reason"), "Publication reason", 5, 500))
        tx.put(pk, "AGENT#" + agent["id"], agent)
        machine = {"sub": "agent_" + agent["id"], "kind": "agent", "agent_id": agent["id"], "agent_revision": agent["revision"],
                   "email": agent["name"], "role": "operator", "status": "active", "joined_at": now()}
        tx.put(pk, "MEMBER#" + machine["sub"], machine)
        audit(tx, pk, actor, "agent_mandate_published", agent)
        return {"agent": agent}

    def _agent_suspend(self, tx, pk, tenant, actor, member, body):
        agent = required(tx, pk, "AGENT#" + identifier(body.get("agent_id")))
        reason = text(body.get("reason"), "Suspension reason", 5, 500)
        agent.update(status="suspended", suspended_at=now())
        tx.put(pk, "AGENT#" + agent["id"], agent)
        machine = tx.get(pk, "MEMBER#agent_" + agent["id"])
        if machine:
            tx.put(pk, "MEMBER#" + machine["sub"], {**machine, "status": "suspended"})
        audit(tx, pk, actor, "agent_suspended", {"id": agent["id"], "reason": reason})
        return {"agent": agent}

    def _agent_key_create(self, tx, pk, tenant, actor, member, body):
        agent = required(tx, pk, "AGENT#" + identifier(body.get("agent_id")))
        if agent["status"] != "active" or agent["config"]["provider"] != "external":
            raise PlatformError("Credentials are issued only for active external agents.", 409, "invalid_state")
        days = body.get("days", 7)
        if type(days) is not int or not 1 <= days <= 90:
            raise PlatformError("Credential lifetime must be 1–90 days.")
        token = "agtu_" + secrets.token_urlsafe(32)
        value = {"id": new_id(), "agent_id": agent["id"], "revision": agent["revision"],
                 "label": text(body.get("label"), "Credential label", 2, 100), "status": "active",
                 "created_by": actor.sub, "created_at": now(), "expires_at": int(time.time()) + days * 86400,
                 "secret_hash": hashlib.sha256(token.encode()).hexdigest()}
        tx.put(pk, credential_key(agent["id"], value["id"]), value, insert_only=True)
        tx.put("AGENTTOKEN#" + value["secret_hash"], "META", {"tenant_id": tenant["id"], "agent_id": agent["id"], "key_id": value["id"]}, insert_only=True)
        public = {k: v for k, v in value.items() if k != "secret_hash"}
        audit(tx, pk, actor, "agent_credential_issued", public)
        return {"credential": public, "agent_token": token}

    def _agent_key_revoke(self, tx, pk, tenant, actor, member, body):
        value = required(tx, pk, credential_key(identifier(body.get("agent_id")), identifier(body.get("key_id"))))
        value.update(status="revoked", revoked_at=now())
        tx.put(pk, credential_key(value["agent_id"], value["id"]), value)
        audit(tx, pk, actor, "agent_credential_revoked", {"id": value["id"], "agent_id": value["agent_id"], "reason": text(body.get("reason"), "Revocation reason", 5, 500)})
        return {"credential": {k: v for k, v in value.items() if k != "secret_hash"}}

    @staticmethod
    def _run_quota(tx, pk, agent):
        day = datetime.now(timezone.utc).date().isoformat()
        key = "AGENTRUNS#" + agent["id"] + "#" + day
        usage = tx.get(pk, key) or {"date": day, "runs": 0}
        if usage["runs"] >= agent["config"]["daily_runs"]:
            raise PlatformError("This agent has reached its daily run budget.", 429, "agent_run_budget")
        usage["runs"] += 1
        tx.put(pk, key, usage)

    def _agent_run(self, tx, pk, tenant, actor, member, body):
        agent = required(tx, pk, "AGENT#" + identifier(body.get("agent_id")))
        if agent["status"] != "active" or tenant["status"] != "active":
            raise PlatformError("The agent and institution must be active.", 409, "invalid_state")
        if agent["config"]["provider"] == "external":
            raise PlatformError("External agents submit through the agent proposal API.")
        if agent["config"]["provider"] == "bedrock" and not self.capabilities()["bedrock"]["configured"]:
            raise PlatformError("The business Bedrock model must be configured before starting model runs.", 503, "provider_unconfigured")
        self._run_quota(tx, pk, agent)
        run = {"id": new_id(), "agent_id": agent["id"], "agent_name": agent["name"], "revision": agent["revision"],
               "provider": agent["config"]["provider"], "status": "queued", "requested_by": actor.sub,
               "created_at": now(), "deadline": int(time.time()) + 900,
               "instruction": text(body.get("instruction", agent["config"]["objective"]), "Run instruction", 10, 1000)}
        run["job_key"] = enqueue(tx, "agent_run", tenant["id"], run["id"])
        tx.put(pk, "RUN#" + run["id"], run, insert_only=True)
        audit(tx, pk, actor, "agent_run_requested", run)
        return {"run": run}

    def _agent_run_cancel(self, tx, pk, tenant, actor, member, body):
        run = required(tx, pk, "RUN#" + identifier(body.get("run_id")))
        if run["status"] not in ("queued", "running"):
            raise PlatformError("This run has finished. Manage its resulting action instead.", 409, "invalid_state")
        if member["role"] not in MANAGERS and run["requested_by"] != actor.sub:
            raise PlatformError("You can cancel only your own run.", 403, "forbidden")
        run.update(status="cancelled", completed_at=now(), reason=text(body.get("reason"), "Cancellation reason", 5, 500))
        tx.put(pk, "RUN#" + run["id"], run)
        tx.delete_work(run["job_key"])
        audit(tx, pk, actor, "agent_run_cancelled", {"id": run["id"], "reason": run["reason"]})
        return {"run": run}

    def submit_agent(self, token, body, request_id):
        if not isinstance(token, str) or not re.fullmatch(r"agtu_[a-zA-Z0-9_-]{43}", token):
            raise PlatformError("A valid agent credential is required.", 401, "agent_unauthenticated")
        if not isinstance(body, dict) or set(body) != {"source_id", "destination_id", "amount", "purpose"}:
            raise PlatformError("Provide only source_id, destination_id, amount and purpose.")
        identifier(request_id, "Idempotency key")
        if len(request_id) < 16:
            raise PlatformError("Use an idempotency key with at least 16 characters.")
        token_hash = hashlib.sha256(token.encode()).hexdigest()
        fingerprint = digest(body)
        def transaction(tx):
            pointer = tx.get("AGENTTOKEN#" + token_hash, "META")
            if not pointer:
                raise PlatformError("Agent credential is unavailable.", 401, "agent_unauthenticated")
            pk = tenant_key(pointer["tenant_id"])
            credential = required(tx, pk, credential_key(pointer["agent_id"], pointer["key_id"]))
            agent = required(tx, pk, "AGENT#" + pointer["agent_id"])
            issuer = tx.get(pk, "MEMBER#" + credential["created_by"])
            issuer_active = issuer and issuer["status"] == "active" and issuer["role"] == "owner"
            if not issuer_active or credential["status"] != "active" or credential["expires_at"] <= time.time() or credential["revision"] != agent["revision"] or agent["status"] != "active":
                raise PlatformError("Agent credential or mandate is expired, revoked or superseded.", 401, "agent_unauthenticated")
            actor = Actor("agent_" + agent["id"], "", False, "agent", credential["id"], credential["created_by"])
            _, tenant, member = access(tx, pointer["tenant_id"], actor, PERMISSIONS["action_propose"])
            key = "AGENTREQUEST#" + agent["id"] + "#" + request_id
            previous = tx.get(pk, key)
            if previous:
                if previous["fingerprint"] != fingerprint:
                    raise PlatformError("Idempotency key was used with different input.", 409, "idempotency_conflict")
                return previous["response"]
            self._run_quota(tx, pk, agent)
            result = self._action_propose(tx, pk, tenant, actor, member, body)
            run = {"id": new_id(), "agent_id": agent["id"], "agent_name": agent["name"], "revision": agent["revision"],
                   "provider": "external", "status": "completed", "created_at": now(), "completed_at": now(),
                   "requested_by": credential["created_by"], "credential_id": credential["id"], "action_id": result["action"]["id"],
                   "decision": result["action"]["status"], "output": body}
            tx.put(pk, "RUN#" + run["id"], run, insert_only=True)
            audit(tx, pk, actor, "external_agent_proposal_recorded", {"run_id": run["id"], "action_id": run["action_id"], "decision": run["decision"]})
            result["run"] = run
            tx.put(pk, key, {"fingerprint": fingerprint, "response": result}, insert_only=True)
            return result
        return self.store.transact(transaction)

    @staticmethod
    def _agent_usage(tx, pk, agent_id, date, code):
        key = f"AGENTUSAGE#{agent_id}#{date}#{code}"
        return key, tx.get(pk, key) or {"date": date, "currency": code, "reserved": 0, "executed": 0}

    def _agent_meter(self, tx, pk, action, date, field, delta):
        if not action.get("agent_id"):
            return
        key, usage = self._agent_usage(tx, pk, action["agent_id"], date, action["currency"])
        usage[field] += delta
        if usage[field] < 0:
            raise PlatformError("Agent usage accounting is inconsistent.", 409, "reservation_inconsistent")
        tx.put(pk, key, usage)
        if field == "reserved":
            destination_key = "AGENTDEST#" + action["agent_id"] + "#" + action["destination_id"]
            incoming = tx.get(pk, destination_key) or {"reserved": 0}
            incoming["reserved"] += delta
            if incoming["reserved"] < 0:
                raise PlatformError("Agent destination reservations are inconsistent.", 409, "reservation_inconsistent")
            tx.put(pk, destination_key, incoming)

    def _agent_checks(self, tx, pk, action, *, held=False):
        if not action.get("agent_id"):
            return []
        agent = required(tx, pk, "AGENT#" + action["agent_id"])
        config = agent["config"]
        date = datetime.now(timezone.utc).date().isoformat()
        _, usage = self._agent_usage(tx, pk, agent["id"], date, action["currency"])
        counted = usage["executed"] + usage["reserved"] - (action["amount"] if held and action["usage_date"] == date else 0)
        valid_credential = True
        if action.get("credential_id"):
            credential = required(tx, pk, credential_key(agent["id"], action["credential_id"]))
            issuer = tx.get(pk, "MEMBER#" + credential["created_by"])
            valid_credential = bool(issuer and issuer["status"] == "active" and issuer["role"] == "owner" and credential["status"] == "active" and credential["expires_at"] > time.time() and credential["revision"] == agent["revision"])
        sponsor = tx.get(pk, "MEMBER#" + (action.get("initiated_by") or ""))
        sponsor_roles = {"owner"} if action.get("credential_id") else self.permissions["agent_run"]
        rules = [
            ("Agent sponsor authorised", bool(sponsor and sponsor["status"] == "active" and sponsor["role"] in sponsor_roles), "Current authority of the human who initiated the run or issued its credential"),
            ("Agent mandate active", agent["status"] == "active" and action["agent_revision"] == agent["revision"], {"status": agent["status"], "revision": agent["revision"]}),
            ("Agent credential valid", valid_credential, "Current credential status and expiry"),
            ("Agent account scope", action["source_id"] in config["source_ids"] and action["destination_id"] in config["destination_ids"], {"sources": config["source_ids"], "destinations": config["destination_ids"]}),
            ("Agent currency scope", action["currency"] in config["currencies"], config["currencies"]),
            ("Agent transaction limit", action["amount"] <= config["transaction_limit"], config["transaction_limit"]),
            ("Agent daily limit", counted + action["amount"] <= config["daily_limit"], {"limit": config["daily_limit"], "committed_and_reserved": counted}),
        ]
        if config["provider"] == "treasury_rule":
            source = required(tx, pk, "ACCOUNT#" + action["source_id"])
            destination = required(tx, pk, "ACCOUNT#" + action["destination_id"])
            incoming = tx.get(pk, "AGENTDEST#" + agent["id"] + "#" + action["destination_id"]) or {"reserved": 0}
            available = source["balance"] - source["reserved"] + (action["amount"] if held else 0)
            committed = destination["balance"] + incoming["reserved"] - (action["amount"] if held else 0)
            rules += [
                ("Agent source floor", available - action["amount"] >= config["source_floor"], {"available": available, "floor": config["source_floor"]}),
                ("Agent reserve target", committed + action["amount"] <= config["destination_target"], {"committed": committed, "target": config["destination_target"]}),
            ]
        return rules

    def _agent_review(self, tx, pk, action):
        return bool(action.get("agent_id") and required(tx, pk, "AGENT#" + action["agent_id"])["config"]["require_review"])
